topology/intel/dmic/dmic-nhlt.c

```diff
@@ -0,0 +1,1827 @@
+
+			val = OUT_GAIN_LEFT_A_GAIN(0);
+			nhlt->dmic_blob_fir[i][di].out_gain_left = val;
+
+			val = OUT_GAIN_RIGHT_A_GAIN(0);
+			nhlt->dmic_blob_fir[i][di].out_gain_right = val;
+
+			/* Write coef RAM A with scaled coefficient in reverse order */
+			length = cfg->fir_a_length;
+			for (j = 0; j < length; j++) {
+				ci = (int32_t)Q_MULTSR_32X32((int64_t)cfg->fir_a->coef[j],
+							     cfg->fir_a_scale, 31,
+							     DMIC_FIR_SCALE_Q, DMIC_HW_FIR_COEF_Q);
+				cu = FIR_COEF_A(ci);
+				/* blob_pdm[i].fir_coeffs[di][j] = cu; */
+				nhlt->dmic_fir_array.fir_coeffs[i][di][j] = cu;
+			}
+			tplg_pp_debug("set fir len a %u and b 0", fir_length);
+			nhlt->dmic_fir_array.fir_len[0] = length;
+			nhlt->dmic_fir_array.fir_len[1] = 0;
+		}
+
+		if (di == 1) {
+			/* FIR B */
+			fir_decim = MAX(cfg->mfir_b - 1, 0);
+			fir_length = MAX(cfg->fir_b_length - 1, 0);
+			val = FIR_CONTROL_B_START(1) |
+				FIR_CONTROL_B_ARRAY_START_EN(array_b) |
+				FIR_CONTROL_B_DCCOMP(dccomp) |
+				FIR_CONTROL_B_MUTE(fir_mute) |
+				FIR_CONTROL_B_STEREO(stereo[i]);
+			nhlt->dmic_blob_fir[i][di].fir_control = val;
+
+			val = FIR_CONFIG_B_FIR_DECIMATION(fir_decim) |
+				FIR_CONFIG_B_FIR_SHIFT(cfg->fir_b_shift) |
+				FIR_CONFIG_B_FIR_LENGTH(fir_length);
+			nhlt->dmic_blob_fir[i][di].fir_config = val;
+			val = DC_OFFSET_LEFT_B_DC_OFFS(DCCOMP_TC0);
+			nhlt->dmic_blob_fir[i][di].dc_offset_left = val;
+
+			val = DC_OFFSET_RIGHT_B_DC_OFFS(DCCOMP_TC0);
+			nhlt->dmic_blob_fir[i][di].dc_offset_right = val;
+
+			val = OUT_GAIN_LEFT_B_GAIN(0);
+			nhlt->dmic_blob_fir[i][di].out_gain_left = val;
+
+			val = OUT_GAIN_RIGHT_B_GAIN(0);
+			nhlt->dmic_blob_fir[i][di].out_gain_right = val;
+
+			/* Write coef RAM B with scaled coefficient in reverse order */
+			length = cfg->fir_b_length;
+			for (j = 0; j < length; j++) {
+				ci = (int32_t)Q_MULTSR_32X32((int64_t)cfg->fir_b->coef[j],
+							     cfg->fir_b_scale, 31,
+							     DMIC_FIR_SCALE_Q, DMIC_HW_FIR_COEF_Q);
+				cu = FIR_COEF_B(ci);
+				/* blob_pdm[i].fir_coeffs[di][j] = cu; */
+				nhlt->dmic_fir_array.fir_coeffs[i][di][j] = cu;
+			}
+			tplg_pp_debug("set fir len b %u", fir_length);
+			nhlt->dmic_fir_array.fir_len[1] = length;
+		}
+	}
+
+	tplg_pp_debug("");
+
+	return 0;
+}
+
+static int dmic_calculate(struct tplg_pre_processor *tplg_pp)
+{
+	struct dmic_calc_matched_modes modes_ab;
+	struct dmic_calc_decim_modes modes_a;
+	struct dmic_calc_decim_modes modes_b;
+	struct dmic_calc_configuration cfg;
+	struct intel_nhlt_params *nhlt;
+	int32_t unmute_ramp_time_ms;
+	int ret = 0;
+	int di;
+
+	nhlt = (struct intel_nhlt_params *)tplg_pp->private_data;
+
+	if (!nhlt)
+		return -EINVAL;
+
+	di = nhlt->dmic_dai_index;
+
+	/* Compute unmute ramp gain update coefficient. Use the value from
+	 * topology if it is non-zero, otherwise use default length.
+	 */
+	/* unmute_ramp related stuff is not used in nhlt at all */
+	if (nhlt->dmic_prm[di].unmute_ramp_time)
```

Comment:
I think you can remove lines 1187 - 1198.

Reply:
ok